### ISSUE TITLE
i want to remove some edges

### ISSUE THREAD

**Issue by pixar0407** [rusty1s/pytorch_geometric#1525]
## ❓ Questions & Help

I wanted to remove some edges from cora dataset.
But i failed to do it. 
here is description.

I am following "introduction by example" of your document(https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html).

So, i am using Cora dataset.
in cora dataset, there are 10,556 edges.

I can see the coo format of edge information like this ( dataset[0].edge_index[0], dataset[0].edge_index[1] )

To change the edge information, I tried like this.(of course, i did the same thing for edge_index[1]. But i omit it for convenience)
code's intention was for getting 9879 edge information out of 10556.

<code starts>
dataset[0].edge_index[0]=torch.as_tensor(dataset[0].edge_index[0].tolist()[-9879:])
<code ends>

but this gave me an error
![111](https://user-images.githubusercontent.com/26530798/89869571-7ed86900-dbef-11ea-9ad5-6930d119df99.JPG)

i think "dataset" has predefined size of tensor.
what i want is just removing the edges. 

Is there any way of changing edge information from dataset?
Or any brilliant hack for bypassing that edge during training and inference?

thanks!

**Comment by treszkai** [comment]
You could remove the edge with a single statement:

```python
data = dataset[0]
data.edge_index = data.edge_index[-9879:]
```

(In case this dataset had edge attributes, you should do the same with `edge_attr`.)

This gives you a single a single `Data` instance. You can't modify the dataset directly -- elements of the dataset are read-only. However, we could do the same in the `pre_transform` of the dataset, which is applied to every datapoint of the dataset (in this case, to the only one):

```python
>>> def pre_transform(d):
  ...:     d.edge_index = d.edge_index[:, -9879:]
  ...:     return d
  ...: 
dataset = CitationFull('temp/Cora', 'Cora', pre_transform=pre_transform)
WARNING:root:The `pre_transform` argument differs from the one used in the pre-processed version of this dataset. If you really want to make use of another pre-processing technique, make sure to delete `temp/Cora/cora/processed` first.
```

As the warning says, delete the "processed" directory first:
```
rm -r temp/Cora/cora/processed
```

Try again:
```
>>> dataset = CitationFull('temp/Cora', 'Cora', pre_transform=pre_transform)
Processing...
Done!
>>> dataset[0].edge_index.shape
torch.Size([2, 9879])
```

**Comment by pixar0407** [comment]
Thank you for replying me.

I didn't know that elements of dataset is read only. It gives me why i got error.
And your suggestion seems pretty good. I think i could use 'pre_transform' in future helpfully.

After I issued my initial question. I had found my way. I will explain this for other people who would suffer this kind of problems. 

I recreated dataset by using 'Data class'.

<code>
from_geometric.data import Data as **dd**

# I transformed original data however i wanted, and made 'o_x', 'o_y', 'o_edge_index'

o_dataset = **dd**(x=o_x, edge_index=o_edge_index, y=o_y)

# now i could use 'o_dataset' as customized dataset
</code>

But in this way, i cannot change the train_mask, test_mask, val_mask, because 'from_geometric.data import Data' does not take 'mask's as its inputs. But in my design, i really don't need to change that.

Thanks for your valuable help!
